openmmtools/integrators.py

```diff
@@ -1895,6 +1895,232 @@ def _add_alchemical_reset_step(self):
         # Add all dependent parameters
         self._add_update_alchemical_parameters_step()
 
+class LoopyAlchemicalNonequilibriumLangevinIntegrator(AlchemicalNonequilibriumLangevinIntegrator):
+    """
+    Subclass of `AlchemicalNonequilibriumLangevinIntegrator` that hardcodes the following integration scheme:
+
+    Step 1: lambda_0 equilibrium sampling
+        run nsteps_eq of BAOAB (V R O R V) integration at the alchemical lambda_0 endstate
+    Step 2: Forward Annealing
+        run n_steps_neq of BAOAB (V R O R V H); 'H': Hamiltonian update step
+    Step 3: lambda_1 equilibrium sampling
+        run Step 1 at the lambda_1 endstate
+    Step 3: Reverse Annealing
+        run Step 2 in the reverse direction
+
+    forward and backward works are saved to `forward_work` and `backward_work`, respectively
+    """
+    def __init__(self,
+                 alchemical_functions = None,
+                 nsteps_eq = 1000,
+                 nsteps_neq = 100,
+                 **kwargs):
+        """
+        arguments
+            nsteps_eq : int
+                number of equilibration steps to run at either endstate
+            nsteps_neq : int
+                number of nonequilibrium annealing steps to run in either direction
+
+        parameters
+            _n_steps_eq : int
+                number of BAOAB loops to run at each endstate in an attempt to generate i.i.d samples from endstates
+        """
+        self._n_steps_eq = nsteps_eq
+        splitting = 'V R O R V'
+        super().__init__(alchemical_functions, splitting, nsteps_neq = nsteps_neq, **kwargs)
+
+    def _add_global_variables(self):
+        """
+        modify the super (_add_global_variables) to add a collector for forward and reverse annealing, as well as to specify the number of equilibration
+        steps at endstates.
+        """
+        super()._add_global_variables()
+        self.addGlobalVariable('forward_work', 0)
+        self.addGlobalVariable('backward_work', 0)
+        self.addGlobalVariable('n_steps_eq', self._n_steps_eq)
+        self.addGlobalVariable('eq_step', 0)
+
+        #overwrite because this is set to 0 in super() (n_H = 1, but this is omitted in the splitting string);
+        #see https://github.com/choderalab/openmmtools/blob/c2b61c410b255c4e08927acf8cfcb1cf46f64b70/openmmtools/integrators.py#L1818
+        self.setGlobalVariableByName('n_lambda_steps', self._n_steps_neq)
+
+    def _add_integrator_steps(self):
+        """
+        hardcode a custom integration scheme specified at the top of the class
+        """
+        #init/reset
+        self.addConstrainPositions() #constrain positions
+        self.addConstrainVelocities() #and velocities
+        self._add_reset_step() #reset lambda, protocol_works, heat, shadow, ghmc, step (unused), lambda_step, and alchemical_params
```

Comment:
add eq_step to docstring, to prevent heart attacks

```suggestion
        self._add_reset_step() #reset lambda, protocol_works, heat, shadow, ghmc, step (unused), lambda_step, eq_step, and alchemical_params
```